packages/flutter/lib/src/widgets/sliver_floating_header.dart

```diff
@@ -0,0 +1,304 @@
+              effectiveScrollOffset = snapAnimation.value;
+              markNeedsLayout();
+            }
+          });
+        snapController!.duration = switch (direction) {
+          ScrollDirection.forward => animationStyle?.duration ?? const Duration(milliseconds: 300),
+          _ => animationStyle?.reverseDuration ?? const Duration(milliseconds: 300),
+        };
+        snapAnimation = snapController!.drive(
+          Tween<double>(
+            begin: effectiveScrollOffset,
+            end: switch (direction) {
+              ScrollDirection.forward => 0,
+              _ => childExtent,
+            },
+          ).chain(
+            CurveTween(
+              curve: switch (direction) {
+                ScrollDirection.forward => animationStyle?.curve ?? Curves.easeInOut,
+                _ => animationStyle?.reverseCurve ?? Curves.easeInOut,
+              }
+            ),
+          ),
+        );
+        snapController!.forward(from: 0.0);
+      }
+    }
+  }
+
+  double get childExtent {
+    if (child == null) {
+      return 0.0;
+    }
+    assert(child!.hasSize);
+    return switch (constraints.axis) {
+      Axis.vertical => child!.size.height,
+      Axis.horizontal => child!.size.width,
+    };
+  }
+
+  @override
+  void detach() {
+    snapController?.dispose();
+    snapController = null; // lazily recreated if we're reattached.
+    super.detach();
+  }
+
+  // True if the header has been laid at at least once (lastScrollOffset != null) and either:
+  // - We're scrolling forward: constraints.scrollOffset < lastScrollOffset
+  // - The header's already partially visible: effectiveScrollOffset < childExtent
+  // Scrolling forwards (towards the scrollable's start) is the trigger that causes the
+  // header to be shown.
+  bool get floatingHeaderNeedsToBeUpdated {
+    return lastScrollOffset != null &&
+      (constraints.scrollOffset < lastScrollOffset! || effectiveScrollOffset < childExtent);
+  }
+
+  @override
+  void performLayout() {
+    if (!floatingHeaderNeedsToBeUpdated) {
+      effectiveScrollOffset = constraints.scrollOffset;
+    } else {
+      double delta = lastScrollOffset! - constraints.scrollOffset; // > 0 when the header is growing
+      if (constraints.userScrollDirection == ScrollDirection.forward) {
+        if (effectiveScrollOffset > childExtent) {
+          effectiveScrollOffset = childExtent; // The header is now just above the start edge of viewport.
+        }
+      } else {
+        // delta > 0 and scrolling forward is a contradiction. Assume that it's noise (set delta to 0).
+        delta = clampDouble(delta, -double.infinity, 0);
+      }
+      effectiveScrollOffset = clampDouble(effectiveScrollOffset - delta, 0.0, constraints.scrollOffset);
+    }
+
+    child?.layout(constraints.asBoxConstraints(), parentUsesSize: true);
+    final double paintExtent = childExtent - effectiveScrollOffset;
+    final double layoutExtent = childExtent - constraints.scrollOffset;
```

Comment:
Yes, although that would cause slightly different behavior when the header animated in our out. If paintExtent and layoutExtent were the same then the list would appear to scroll up or down when the header snapped in or out of view. That's a useful alternative (that's the way the Instagram floating header works) although it's not the same way that the existing floating SliverPersistentHeader works. 

It would be easy enough to make this an option although, if it's OK with you, I'll do that in a separate PR.

Reply:
Ah got it! I agree to be consistent with SliverPersistentHeader:) Thanks for the explanation!